controller/namespace/namespacecontroller.go

```diff
@@ -16,139 +16,106 @@ limitations under the License.
+// PostgreSQL Operator are added and deleted.
+func NewNamespaceController(controllerManager controller.ManagerInterface,
+	informer coreinformers.NamespaceInformer) (*Controller, error) {
 
-	err := c.watchNamespaces(c.Ctx)
-	if err != nil {
-		log.Errorf("Failed to register watch for namespace resource: %v", err)
-		return err
+	controller := &Controller{
+		ControllerManager: controllerManager,
+		Informer:          informer,
 	}
 
-	<-c.Ctx.Done()
-	return c.Ctx.Err()
+	return controller, nil
 }
 
-// watchNamespaces is the event loop for namespace resources
-func (c *Controller) watchNamespaces(ctx context.Context) error {
-	log.Info("starting namespace controller")
-
-	//watch all namespaces
-	ns := ""
-
-	source := cache.NewListWatchFromClient(
-		c.NamespaceClientset.CoreV1().RESTClient(),
-		"namespaces",
-		ns,
-		fields.Everything())
-
-	_, controller := cache.NewInformer(
-		source,
-
-		// The object type.
-		&v1.Namespace{},
-
-		// resyncPeriod
-		// Every resyncPeriod, all resources in the cache will retrigger events.
-		// Set to 0 to disable the resync.
-		0,
+// AddNamespaceEventHandler adds the pod event handler to the namespace informer
+func (c *Controller) AddNamespaceEventHandler() {
 
-		// Your custom resource event handlers.
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    c.onAdd,
-			UpdateFunc: c.onUpdate,
-			DeleteFunc: c.onDelete,
-		})
+	c.Informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
 
-	go controller.Run(ctx.Done())
-
-	return nil
+	log.Debugf("Namespace Controller: added event handler to informer")
 }
 
+// onUpdate is called when a namespace is added
 func (c *Controller) onAdd(obj interface{}) {
+
 	newNs := obj.(*v1.Namespace)
 
 	log.Debugf("[namespace Controller] OnAdd ns=%s", newNs.ObjectMeta.SelfLink)
 	labels := newNs.GetObjectMeta().GetLabels()
 	if labels[config.LABEL_VENDOR] != config.LABEL_CRUNCHY || labels[config.LABEL_PGO_INSTALLATION_NAME] != operator.InstallationName {
 		log.Debugf("namespace Controller: onAdd skipping namespace that is not crunchydata or not belonging to this Operator installation %s", newNs.ObjectMeta.SelfLink)
 		return
-	} else {
-		log.Debugf("namespace Controller: onAdd crunchy namespace %s created", newNs.ObjectMeta.SelfLink)
-		c.PodController.SetupWatch(newNs.Name)
-		c.JobController.SetupWatch(newNs.Name)
-		c.PgpolicyController.SetupWatch(newNs.Name)
-		c.PgreplicaController.SetupWatch(newNs.Name)
-		c.PgclusterController.SetupWatch(newNs.Name)
-		c.PgtaskController.SetupWatch(newNs.Name)
 	}
 
+	log.Debugf("namespace Controller: onAdd crunchy namespace %s created", newNs.ObjectMeta.SelfLink)
+	c.ControllerManager.AddAndRunControllerGroup(newNs.Name)
 }
 
-// onUpdate is called when a pgcluster is updated
+// onUpdate is called when a namespace is updated
 func (c *Controller) onUpdate(oldObj, newObj interface{}) {
-	//oldNs := oldObj.(*v1.Namespace)
```

Comment:
What about when `oldNs` was being watched but `newNs` wouldn't? Seems like we should `RemoveGroup()` in that case.